File: algo/que/que.py
from collections import deque
from typing import List


class Que:
    def __init__(self):
        self.que = []

    @property
    def get_que(self):
        return self.que

    def enqueue(self, data):
        self.que.append(data)

    def deque(self):
        self.que.pop(0)

    def que_reverse(self, queue: list[int]):
        new_que = []
        while queue:
            new_que.append(queue.pop())

        return new_que


if __name__ == "__main__":
    # q = deque()
    # q.append(1)
    # q.append(2)
    # q.append(3)
    # q.append(4)
    # q.append(5)
    # print(q)
    # q.popleft()
    # q.popleft()
    # q.popleft()
    # print(q)

    q = Que()
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    q.enqueue(4)
    q.enqueue(5)
    print(q.get_que)
    q.deque()
    q.deque()
    q.deque()
    print(q.get_que)

    print(q.que_reverse(q.get_que))

File: algo/que/que.js
class Queue {
  constructor(){
    this.que = []
  }

  get getQueue() {
    return this.que
  }

  enqueue(data) {
    this.que.push(data)
  }

  dequeue() {
    if (this.que.length > 0){
      return this.que.shift()
    }
    return undefined
  }

  reverseQueue(arr){
    const newQue = []

    const tempArr = [...arr];
    while (tempArr.length > 0) {
      newQue.push(tempArr.pop());
    }
    return newQue;
  }
}

const q = new Queue()
q.enqueue(1);
q.enqueue(2);
q.enqueue(3);

console.log(q.getQueue); // [1, 2, 3]

console.log(q.dequeue()); // 1
console.log(q.dequeue()); // 2
console.log(q.getQueue); // [2, 3]

const originalArray = [5, 6, 7, 8];
const reversedArray = q.reverseQueue(originalArray);
console.log(reversedArray); // [8, 7, 6, 5]
console.log(originalArray); // [5, 6, 7, 8] (元の配列は変更されない)
